core/field_number.js

```diff
@@ -51,316 +45,267 @@ goog.require('goog.userAgent');
+  /**
+   * Call for when the num-pad erase button is touched.
+   * Determine what the user is asking to erase, and erase it.
+   * @param {Event} e DOM event triggering the touch.
+   */
+  numPadEraseButtonTouch(e) {
+    // Old value of the text field
+    var oldValue = this.htmlInput_.value;
+    // Determine what is selected to erase (if anything)
+    var selectionStart = this.htmlInput_.selectionStart;
+    var selectionEnd = this.htmlInput_.selectionEnd;
+
+    // If selection is zero-length, shift start to the left 1 character
+    if (selectionStart == selectionEnd) {
+      selectionStart = Math.max(0, selectionStart - 1);
+    }
 
-  // Prevent default to not lose input focus
-  e.preventDefault();
-};
+    // Cut out selected range
+    var newValue = oldValue.slice(0, selectionStart) +
+        oldValue.slice(selectionEnd);
 
-/**
- * Call for when the num-pad erase button is touched.
- * Determine what the user is asking to erase, and erase it.
- * @param {Event} e DOM event triggering the touch.
- */
-Blockly.FieldNumber.numPadEraseButtonTouch = function(e) {
-  // Old value of the text field
-  var oldValue = Blockly.FieldTextInput.htmlInput_.value;
-  // Determine what is selected to erase (if anything)
-  var selectionStart = Blockly.FieldTextInput.htmlInput_.selectionStart;
-  var selectionEnd = Blockly.FieldTextInput.htmlInput_.selectionEnd;
-
-  // If selection is zero-length, shift start to the left 1 character
-  if (selectionStart == selectionEnd) {
-    selectionStart = Math.max(0, selectionStart - 1);
-  }
+    this.updateDisplay_(newValue, selectionStart);
 
-  // Cut out selected range
-  var newValue = oldValue.slice(0, selectionStart) +
-      oldValue.slice(selectionEnd);
+    // This is just a click.
+    Blockly.Touch.clearTouchIdentifier();
 
-  Blockly.FieldNumber.updateDisplay_(newValue, selectionStart);
+    // Prevent default to not lose input focus which resets cursors in Chrome
+    e.preventDefault();
+  }
 
-  // This is just a click.
-  Blockly.Touch.clearTouchIdentifier();
+  doClassValidation_(
+    newValue
+  ) {
+    if (newValue === '') return newValue;
+    return super.doClassValidation_(newValue);
+  }
 
-  // Prevent default to not lose input focus which resets cursors in Chrome
-  e.preventDefault();
-};
+  /**
+   * Update the displayed value and resize/scroll the text field as needed.
+   * @param {string} newValue The new text to display.
+   * @param {string} newSelection The new index to put the cursor
+   * @private.
+   */
+  updateDisplay_(newValue, newSelection) {
+    var htmlInput = this.htmlInput_;
+    // Updates the display. The actual setValue occurs when editing ends.
+    htmlInput.value = newValue;
+    this.setValue(newValue);
+    // Resize and scroll the text field appropriately
+    super.resizeEditor_();
+    htmlInput.setSelectionRange(newSelection, newSelection);
+    htmlInput.scrollLeft = htmlInput.scrollWidth;
+    // this.validate_();
```

Comment:
I may not be reading this correctly but my instinct is you shouldn't need to set the htmlInput value since we [do that in core](https://github.com/google/blockly/blob/185c82a991484e68b1a76825f2c4b082b0168677/core/field_input.ts#L608). You just need to call `setEditorValue_` instead of `setValue`.

Also find the call to `super.resizeEditor` pretty suss. I'd expect that to just happen during rendering and you wouldn't have to do it here? Can you try removing it to see what it's doing?

Reply:
Done.